File: pptx-service/renderers/cover.py
from builder import add_bg, add_textbox, add_rect, add_eyebrow
from theme import ThemeColors


def render(slide, data: dict, t: ThemeColors) -> None:
    add_bg(slide, t.bg)
    title = data.get("title", "")
    subtitle = data.get("subtitle", "")
    eyebrow = data.get("eyebrow", "")

    add_rect(slide, 0.6, 1.8, 0.08, 3.8, fill_hex=t.accent)

    if eyebrow:
        add_eyebrow(slide, eyebrow, t, y=1.7)

    add_textbox(slide, 0.83, 2.2, 11.0, 3.5, title,
                t.font_display, 80, t.text, bold=True, wrap=True)

    if subtitle:
        add_textbox(slide, 0.83, 5.9, 9.0, 0.8, subtitle,
                    t.font_body, 28, t.muted)

File: lib/pptx/slides/cover.ts
import type { CoverSlide } from '@/lib/types';
import {
  px, py, SLIDE_W,
  setBackground, addRect, addTextBox, addPageNumber, addEyebrow,
  type PptxTheme,
} from '../helpers';
import type PptxGenJS from 'pptxgenjs';
type PSlide = ReturnType<InstanceType<typeof PptxGenJS>['addSlide']>;

export function render(slide: PSlide, data: CoverSlide, t: PptxTheme, n: number, total: number): void {
  setBackground(slide, '#' + t.bg);

  // Vertical accent bar
  addRect(slide, t.paddingIn - px(24), py(300), px(8), py(300), '#' + t.accent);

  if (data.eyebrow) addEyebrow(slide, data.eyebrow, t);

  const titleY = py(340);
  const titleW = SLIDE_W - 2 * t.paddingIn;

  if (data.highlight) {
    const idx = data.title.indexOf(data.highlight);
    if (idx >= 0) {
      const before = data.title.slice(0, idx);
      const after = data.title.slice(idx + data.highlight.length);
      const runs: object[] = [];
      if (before) runs.push({ text: before, options: { color: t.text, fontFace: t.fontDisplay, bold: true } });
      runs.push({ text: data.highlight, options: { color: t.accent, fontFace: t.fontDisplay, bold: true } });
      if (after) runs.push({ text: after, options: { color: t.text, fontFace: t.fontDisplay, bold: true } });
      slide.addText(runs as Parameters<PSlide['addText']>[0], {
        x: t.paddingIn, y: titleY, w: titleW, fontSize: t.heroSize, bold: true, wrap: true,
      });
    } else {
      addTextBox(slide, data.title, { x: t.paddingIn, y: titleY, w: titleW, fontSize: t.heroSize, color: '#' + t.text, fontFace: t.fontDisplay, bold: true });
    }
  } else {
    addTextBox(slide, data.title, { x: t.paddingIn, y: titleY, w: titleW, fontSize: t.heroSize, color: '#' + t.text, fontFace: t.fontDisplay, bold: true });
  }

  if (data.subtitle) {
    addTextBox(slide, data.subtitle, {
      x: t.paddingIn, y: py(530), w: SLIDE_W - 2 * t.paddingIn,
      fontSize: t.bodySize, color: '#' + t.muted, fontFace: t.fontBody,
    });
  }

  addPageNumber(slide, n, total, t);
}
